subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/ValidationActions.java

```diff
@@ -183,15 +184,20 @@ private static void reportFileTreeWithFileRoot(String propertyName, TaskValidati
         );
     }
 
-    private static void reportCannotWriteToFile(String propertyName, TaskValidationContext context, String cause) {
-        context.visitPropertyProblem(problem ->
-            problem.withId(ValidationProblemId.CANNOT_WRITE_OUTPUT)
-                .reportAs(ERROR)
-                .forProperty(propertyName)
-                .withDescription(() -> "is not writable because " + cause)
-                .happensBecause(() -> "Cannot write a file to a location pointing at a directory")
-                .addPossibleSolution(() -> "Configure '" + propertyName + "' to point to a file, not a directory")
-                .documentedAt("validation_problems", "cannot_write_output")
+    private static void reportCannotWriteToFile(String propertyName, TaskValidationContext context, String cause, boolean ancestorFile) {
+        context.visitPropertyProblem(problem -> {
+                PropertyProblemBuilder problemBuilder = problem.withId(ValidationProblemId.CANNOT_WRITE_OUTPUT)
+                    .reportAs(ERROR)
+                    .forProperty(propertyName)
+                    .withDescription(() -> "is not writable because " + cause)
+                    .happensBecause(() -> "Cannot write a file to a location pointing at a directory")
+                    .addPossibleSolution(() -> "Configure '" + propertyName + "' to point to a file, not a directory")
```

Comment:
This message is not helpful if `ancestorFile == true`, no? I mean, that's the case when you point the output to `.../src/main/java`, and `.../src/main` turns out to be a file. So suggesting to point the output to something that's not a directory is not helping, because that's not the problem. We should provide an alternative message here instead.

Reply:
Whoops, sorry that I missed the response here.
I fully agree, I just refrained from changing the existing solution as I might not have full insight and the one invented it could have seen a use-case I don't .
But I fully agree with you.

What solution text do you suggest for the ancestor file case?

Reply:
Maybe like this:
```java
private static void reportCannotWriteFileToDirectory(String propertyName, TaskValidationContext context, File file) {
    context.visitPropertyProblem(problem -> {
            PropertyProblemBuilder problemBuilder = problem.withId(ValidationProblemId.CANNOT_WRITE_OUTPUT)
                .reportAs(ERROR)
                .forProperty(propertyName)
                .withDescription(() -> "is not writable because '" + file + "' is not a file")
                .happensBecause(() -> "Cannot write a file to a location pointing at a directory")
                .addPossibleSolution(() -> "Configure '" + propertyName + "' to point to a file, not a directory")
                .addPossibleSolution(() -> "Annotate '" + propertyName + "' with @OutputDirectory instead of @OutputFiles")
                .documentedAt("validation_problems", "cannot_write_output");
        }
    );
}

private static void reportCannotCreateParentDirectories(String propertyName, TaskValidationContext context, File file, File ancestor) {
    context.visitPropertyProblem(problem -> {
            PropertyProblemBuilder problemBuilder = problem.withId(ValidationProblemId.CANNOT_WRITE_OUTPUT)
                .reportAs(ERROR)
                .forProperty(propertyName)
                .withDescription(() -> "is not writable because '" + file + "' ancestor '" + ancestor + "' is not a directory")
                .happensBecause(() -> "Cannot create parent directories that are existing as file")
                .addPossibleSolution(() -> "Configure '" + propertyName + "' to point to the correct location")
                .documentedAt("validation_problems", "cannot_write_output");
        }
    );
}
```
?

Reply:
I pushed a new and rebased version